### PR TITLE
Added Endpoints for Points

### PR DESCRIPTION
**Pull request by mubashir494**
Closes : #361 

Tasks Done:
- Added Enpoints to Gift and confirm Points
- Added API routes for waller
- Added Corrosponding tests

Tasks TBD:
- Add some Analytics

### CODE REVIEW

**Review by mathieutrudeau — CHANGES_REQUESTED** [review]
There are conflicts with the main branch that have to be resolved before merging.

File: Backend/Backend/Services/StripeServices.cs [hunk 1]
```diff
@@ -0,0 +1,56 @@
...
+    {
+        private readonly SessionService _sessionService;
+        private readonly IConfiguration _configuration;
+        public StripeServices(SessionService sessionService, IConfiguration configuration)
+        {
+            _sessionService = sessionService; 
+            _configuration = configuration;
+
+        }
+        public async Task<string> CreatePaymentSession(int points,Guid pointId)
```
Comment (mathieutrudeau) [inline Backend/Backend/Services/StripeServices.cs:19]: A lot of hardcoded values. Might want to create constants for them.

**Review by mathieutrudeau — APPROVED** [review]

File: Backend/Backend.Tests/WalletTests.cs [hunk 2]
```diff
@@ -0,0 +1,226 @@
...
+            bool response = false;
+            try
+            {   
+                // Try to Withdraw 
+                response = _walletServices.Withdraw(_user.Object.First().Id, 5).Result;
+               
+            }
+            catch (Exception e)
+            {
+                // Else Assert fail
```
Comment (BigSaladMan) [inline Backend/Backend.Tests/WalletTests.cs:53]: You dont need to change this, but you dont need to have THAT many comments haha

File: Backend/Backend/Controllers/Responses/StripeResponse.cs [hunk 3]
```diff
@@ -0,0 +1,6 @@
+﻿namespace Backend.Controllers.Responses
+{
+    public class StripeResponse
```
Comment (BigSaladMan) [inline Backend/Backend/Controllers/Responses/StripeResponse.cs:3]: What is the point of this class?

File: Backend/Backend/Services/SocialService.cs [hunk 4]
```diff
@@ -482,4 +501,91 @@ public async Task<bool> RemoveReviewFromPodcastAsync(Guid podcastId, User user)
...
+
+        }
+        // Limit the points to 200
+        if(points > 200)
+        {
+            throw new Exception("You cant gift more then 200 points");
+        }
+
+        // Check whether the person gifting is different person not the podcaster
+        Episode? ep1 = await _db.Episodes.Include(u => u.Podcast).FirstOrDefaultAsync(u => u.Podcast.PodcasterId == user.Id);
```
Comment (BigSaladMan) [inline Backend/Backend/Services/SocialService.cs:532]: I think you mightve done this check wrong. 
Would this not just check if the guy is a podcaster in general, not the owner of this particular podcast?
I think this wouldve been the solution, unless im mistaken here

 Episode? ep1 = await _db.Episodes.Include(u => u.Id == episodeId).FirstOrDefaultAsync(u => u.Podcast.PodcasterId == user.Id);

File: Backend/Backend/Services/SocialService.cs [hunk 5]
```diff
@@ -482,4 +501,91 @@ public async Task<bool> RemoveReviewFromPodcastAsync(Guid podcastId, User user)
...
+
+        // if he is a podcaster then throw an error
+        if(ep1 != null)
```
Comment (BigSaladMan) [inline Backend/Backend/Services/SocialService.cs:535]: Then, if you changed it above, this check should be different

File: Backend/Backend/Services/SocialService.cs [hunk 6]
```diff
@@ -482,4 +501,91 @@ public async Task<bool> RemoveReviewFromPodcastAsync(Guid podcastId, User user)
...
+        {
+            throw new Exception("You cant gift points to yourself");
+
+        }
+
+        // If all is good then add points to the DB
+        Guid PointId = Guid.NewGuid();
+        
+        await _db.Points.AddAsync(new Points { Id = PointId, EpisodeId = episodeId, UserId = user.Id, PointCount = points,CreatedAt = DateTime.Now, Amount = points * POINT_CONVERSION });
```
Comment (BigSaladMan) [inline Backend/Backend/Services/SocialService.cs:544]: wait does this mean that when you gift points, they only get a tenth of it?!?!

File: Backend/Backend/Services/StripeServices.cs [hunk 7]
```diff
@@ -0,0 +1,56 @@
...
+            meta.Add("id", pointId.ToString());
+            var options = new SessionCreateOptions()
+            {
+
+                PaymentMethodTypes = new List<string> { "card" },
+                LineItems = new List<SessionLineItemOptions> { new SessionLineItemOptions
+                {
+                    PriceData = new SessionLineItemPriceDataOptions()
+                    {
+                        Currency = "cad",
```
Comment (BigSaladMan) [inline Backend/Backend/Services/StripeServices.cs:31]: Might want to create an enum with the accepted currency types

File: Backend/Backend/Services/WalletServices.cs [hunk 8]
```diff
@@ -0,0 +1,101 @@
...
+
+        }
+        /// <summary>
+        /// Gets the current Balance of the User
+        /// </summary>
+        /// <param name="userId"></param>
+        /// <returns></returns>
+        public async Task<double> GetUserBalance(Guid userId)
+        {
+            return await _db.Transactions.Where(u => u.UserId == userId).SumAsync(u => u.Amount);
```
Comment (BigSaladMan) [inline Backend/Backend/Services/WalletServices.cs:59]: wait, why not save the current balance of the user on the user itself? this feels REALLY convoluted

File: Backend/Backend/Services/WalletServices.cs [hunk 9]
```diff
@@ -0,0 +1,101 @@
...
+        /// Gets the last 5 Days Balance
+        /// </summary>
+        /// <param name="userId"></param>
+        /// <returns></returns>
+        public async Task<List<Activity>> Last5DaysBalance(Guid userId)
+        {
+
+            List<int> numbers = new List<int>() { 0, 1, 2, 3, 4 };
+            List<Activity> activty = new List<Activity>();
+            foreach (var number in numbers)
```
Comment (BigSaladMan) [inline Backend/Backend/Services/WalletServices.cs:89]: Bro, just use a regular for loop 
then
for(int i =0; i<=4; i--;)

**Review by BigSaladMan — APPROVED** [review]
A bunch of small things you should probably fix, but nothing really game breaking

File: Backend/Backend/Services/SocialService.cs (hunk 6 above)
Comment (mubashir494) [reply Backend/Backend/Services/SocialService.cs:544]: So each point is 0.1 CAD so if you gift 10 points then its 1CAD

File: Backend/Backend/Services/SocialService.cs (hunk 4 above)
Comment (mubashir494) [reply Backend/Backend/Services/SocialService.cs:532]: Yeah you are right i will resolve this is in my commit

**Pull request merged**